torch/utils/checkpoint.py

```diff
@@ -1181,133 +1173,233 @@ def _detach(x):
-        >>>     return _pt2_selective_checkpoint_context_fn_gen(get_custom_policy())
+        >>> def policy_fn(ctx, op, *args, **kwargs):
+        >>>    if op in ops_to_save:
+        >>>        return CheckpointPolicy.SAVE_NON_OVERRIDABLE
+        >>>    else:
+        >>>        return CheckpointPolicy.RECOMPUTE_OVERRIDABLE
         >>>
-        >>> def gn(x, y):
-        >>>     return torch.sigmoid(torch.matmul(torch.matmul(x, y), y)) * y
+        >>> context_fn = gen_selective_checkpoint_context_fn(policy_fn)
         >>>
-        >>> def fn(x, y):
-        >>>     return torch.utils.checkpoint.checkpoint(
-        >>>         gn, x, y,
-        >>>         use_reentrant=False,
-        >>>         context_fn=selective_checkpointing_context_fn,
-        >>>     )
+        >>> # or equivalently
+        >>> context_fn = gen_selective_checkpoint_context_fn(ops_to_save)
         >>>
-        >>> x = torch.randn(4, 4, requires_grad=True)
-        >>> y = torch.randn(4, 4, requires_grad=True)
+        >>> def fn(x, y):
+        >>>     return torch.sigmoid(torch.matmul(torch.matmul(x, y), y)) * y
         >>>
-        >>> compiled_fn = torch.compile(fn)
+        >>> out = torch.utils.checkpoint.checkpoint(
+        >>>     fn, x, y,
+        >>>     use_reentrant=False,
+        >>>     context_fn=context_fn,
+        >>> )
     """
-    storage: Dict[Any, List[Any]] = defaultdict(list)
-    return _CachingTorchDispatchMode(policy_fn, storage), _CachedTorchDispatchMode(policy_fn, storage)
+    # NB: If grad_mode is disabled, checkpoint would not run forward under
+    #     context_fn anyway, so proceed as usual.
+    if isinstance(policy_fn_or_list, list):
+        for op in policy_fn_or_list:
+            if not isinstance(op, torch._ops.OpOverload):
+                _extra_msg = (
+                    "Please update the OpOverloadPacket to a specific OpOverload."
+                    "For example, if you have `torch.ops.aten.mm`, change it to `torch.ops.aten.mm.default`."
+                ) if isinstance(op, torch._ops.OpOverloadPacket) else ""
+                raise ValueError(
+                    f"Expected op in `op_list` to be an OpOverload but got: {op} "
+                    f"of type {type(op)}. {_extra_msg}"
+                )
+
+        def policy_fn(ctx, op, *args, **kwargs):
+            if op in policy_fn_or_list:
+                return CheckpointPolicy.SAVE_NON_OVERRIDABLE
+            else:
+                return CheckpointPolicy.RECOMPUTE_OVERRIDABLE
+    elif callable(policy_fn_or_list):
+        policy_fn = policy_fn_or_list
+    else:
+        raise TypeError("policy_fn_or_list must be either a function or a list of ops.")
 
+    def context_fn():
+        storage: Dict[Any, List[Any]] = defaultdict(list)
+        # TODO: when would I need to deepcopy the policy_fn?
+        return (
+            _CachingTorchDispatchMode(policy_fn, storage),
+            _CachedTorchDispatchMode(policy_fn, storage),
+        )
```

Comment:
The `deepcopy` for the policy function can be useful if the policy is not a function, but a class that holds state.

For example, imagine that you have a policy that saves the second matmul. It could look something like
```python
class Policy:
    def __init__(self):
        self.counter = 0

    def __call__(self, mode, func, *args, **kwargs):
            counter = self.counter
            self.counter += 1
            if counter == 1 and func is torch.ops.aten.mm.default:
                return CheckpointPolicy.SAVE_NON_OVERRIDABLE
            return CheckpointPolicy.RECOMPUTE_OVERRIDABLE
                
```

if we don't perform the deepcopy, the `counter` will be shared between the `Caching` and the `Cached` dispatch modes, which will not behave as expected.

This type of pattern is what I used for the [automatic policy selection in xFormers](https://github.com/facebookresearch/xformers/blob/ea4a5508879d31424be5e9b30747fbd162b778a3/xformers/checkpoint.py#L450-L461), which requires finer-grained control of the operations being used, including which operation is used (first, second, etc).

Maybe there is a better way of doing this (like using the `mode` argument to reset the counter), but the deepcopy was easy enough back then (when I didn't have the `mode` argument in the policy, which btw I don't yet use and I'm not sure it's super needed)

Reply:
I see, thanks for the context!

> Maybe there is a better way of doing this (like using the mode argument to reset the counter), but the deepcopy was easy enough back then (when I didn't have the mode argument in the policy, which btw I don't yet use and I'm not sure it's super needed)

I think I prefer this idea actually because it may be possible for some one to wish to share some state during the forward and recompute for whatever reason (I don't know any concrete use cases yet). 

One would need to write policy_fn a little differently though. (Hopefully straighforward if we document that a single instance of policy_fn is shared between forward and recompute)
```python
class Policy:
    def __init__(self):
        self.counter = 0
        self.recompute_counter = 0

    def __call__(self, ctx, func, *args, **kwargs):
            counter = self.recompute_counter if ctx.is_recompute() else self.counter
            counter += 1
            if counter == 1 and func is torch.ops.aten.mm.default:
                return CheckpointPolicy.MUST_SAVE
            return CheckpointPolicy.PREFER_RECOMPUTE
```